pulp_ansible/app/models.py

```diff
@@ -275,6 +313,368 @@ class Meta:
+        log.info(f"Transparency log entry created at index: {result.log_entry.log_index}")
+        signing_result["signature"] = result.b64_signature
+        signing_result["certificate"] = result.cert_pem
+        signing_result["bundle"] = result._to_bundle().to_json()
+
+        return signing_result
+
+    async def sigstore_asign(self, input_digest, private_key, cert):
+        """Sign collections with Sigstore asynchronously."""
+        b64_cert = base64.b64encode(cert.public_bytes(encoding=serialization.Encoding.PEM))
+        signing_result = {}
+        artifact_signature = await sync_to_async(private_key.sign)(
+            input_digest, ec.ECDSA(Prehashed(hashes.SHA256()))
+        )
+        b64_artifact_signature = base64.b64encode(artifact_signature).decode()
+        rekor_post_entry_payload = {
+            "kind": "hashedrekord",
+            "apiVersion": "0.0.1",
+            "spec": {
+                "signature": {
+                    "content": b64_artifact_signature,
+                    "publicKey": {"content": b64_cert.decode()},
+                },
+                "data": {"hash": {"algorithm": "sha256", "value": input_digest.hex()}},
+            },
+        }
+
+        rekor_post_entries_url = urljoin(self.rekor.url, "log/entries/")
+
+        async with aiohttp.ClientSession(
+            headers={"Content-Type": "application/json", "Accept": "application/json"}
+        ) as session:
+            try:
+                async with session.post(
+                    rekor_post_entries_url, json=rekor_post_entry_payload
+                ) as resp:
+                    rekor_response = await resp.json()
+            except aiohttp.web.HTTPError as http_error:
+                raise RekorClientError from http_error
+
+        log_entry = LogEntry._from_response(rekor_response)
+        log.info(f"Transparency log entry created with index: {log_entry.log_index}")
+
+        result = SigningResult(
+            input_digest=input_digest.hex(),
+            cert_pem=PEMCert(cert.public_bytes(encoding=serialization.Encoding.PEM).decode()),
+            b64_signature=b64_artifact_signature,
+            log_entry=log_entry,
+        )
+
+        signing_result["signature"] = result.b64_signature
+        signing_result["certificate"] = result.cert_pem
+        signing_result["bundle"] = result._to_bundle().to_json()
+
+        return signing_result
+
+    def save(self, *args, **kwargs):
+        """Override the base `save` method to properly format PEM-encoded files."""
+        if self.rekor_root_pubkey:
+            self.rekor_root_pubkey = validate_and_format_pem_public_key(self.rekor_root_pubkey)
+        if self.ctfe_pubkey:
+            self.ctfe_pubkey = validate_and_format_pem_public_key(self.ctfe_pubkey)
+        super(SigstoreSigningService, self).save(*args, **kwargs)
+
+    class Meta:
+        default_related_name = "%(app_label)s_%(model_name)s"
+
+
+class SigstoreVerifyingService(BaseModel):
```

Comment:
I remember this was the approach I went with in my first version of this PR, but during a discussion we decided it would be better to separate signing and verifying capabilities for repositories. Is it still the case?

Reply:
No it still is the case that they should be separate. I'm interested to see the changes to this model with the sigstore-bundle changes in the new V2.

Reply:
Remember everyone can verify signatures all the time. But signing needs special setup. So there is no merging.
Also if we need something like a verifying service, the crypto system is inherently broken. The signature itself must contain all the information needed to verify it. And the public keys used for verifying are (as the name suggests) public. Please do not make verifying harder than it is.
The signing service is designed that way to allow using secret keys that pulp will never even touch. For verifying, this does not apply.

Reply:
I understand your concern, but here you need information about the Rekor instance that was used to log the signature to retrieve an inclusion proof from the transparency log (which is the stronger kind of verification, even if offline verification is also possible), and so you need to keep a number of parameters related to signing available to verifiers, like the Rekor instance URL and public key for instance, hence the need for a "verifying service".

Reply:
The only thing we need to do is tell pulp which entities to trust for a given repository. Again no secret bits involved. This is (at best) the equivalent of a public gpg key. Maybe it needs a separate db-table, because somehow sigstore defines complex as the new simple. But it is not going to reflect the modus operandi of a signing service. So no sevice and no setup via pulpcore-manager. If the user needs the admin to tell their repository which sigs to trust, this make for a very bad user experience.